packages/tpetra/core/example/Finite-Element-Assembly/fem_assembly_type3_staticprofile.cpp

```diff
@@ -0,0 +1,322 @@
+
+  // Loop over owned elements:
+  for(size_t element_gidx=0; element_gidx<mesh.getNumOwnedElements(); element_gidx++)
+  {
+    // Get the stiffness matrix for this element
+    ReferenceQuad4(element_matrix);
+
+    // Fill the global column ids array for this element
+    for(size_t element_node_idx=0; element_node_idx<owned_element_to_node_ids.extent(1); element_node_idx++)
+    {
+      column_global_ids[element_node_idx] = owned_element_to_node_ids(element_gidx, element_node_idx);
+    }
+    
+    // For each node (row) on the current element:
+    // - populate the values array
+    // - add values to crs_matrix if the row is owned.
+    //   Note: hardcoded 4 here because we're using quads.
+    for(size_t element_node_idx=0; element_node_idx<4; element_node_idx++)
+    { 
+      GlobalOrdinal global_row_id = owned_element_to_node_ids(element_gidx, element_node_idx);
+      if(mesh.nodeIsOwned(global_row_id)) 
+      {
+        for(size_t col_idx=0; col_idx<4; col_idx++)
+        {
+          column_scalar_values[col_idx] = element_matrix(element_node_idx, col_idx);
+        }
+        crs_matrix->sumIntoGlobalValues(global_row_id, column_global_ids, column_scalar_values);
+      }
+    }
+  }
+
+  // Loop over ghost elements:
+  // - This loop is the same as the element loop for owned elements, but this one
+  //   is for ghost elements.
+  for(size_t element_gidx=0; element_gidx<mesh.getNumGhostElements(); element_gidx++)
+  {
+    ReferenceQuad4(element_matrix);
+
+    for(size_t element_node_idx=0; element_node_idx<ghost_element_to_node_ids.extent(1); element_node_idx++)
+    {
+      column_global_ids[element_node_idx] = ghost_element_to_node_ids(element_gidx, element_node_idx);
+    }
+    
+    for(size_t element_node_idx=0; element_node_idx<4; element_node_idx++)
+    { 
+      GlobalOrdinal global_row_id = ghost_element_to_node_ids(element_gidx, element_node_idx);
+      if(mesh.nodeIsOwned(global_row_id)) 
+      {
+        for(size_t col_idx=0; col_idx<4; col_idx++)
+        {
+          column_scalar_values[col_idx] = element_matrix(element_node_idx, col_idx);
+        }
+        crs_matrix->sumIntoGlobalValues(global_row_id, column_global_ids, column_scalar_values);
+      }
+    }
+  }
+  timerElementLoopMatrix = Teuchos::null;
+
+  // After the contributions are added, 'finalize' the matrix using fillComplete()
+  {
+    RCP<TimeMonitor> timerFillCompleteMatrix = rcp(new TimeMonitor(*TimeMonitor::getNewTimer("4) FillComplete (Matrix)")));
+    crs_matrix->fillComplete();
+  }
+
+  // Print out crs_matrix details.
+  #if PRINT_VERBOSE
+  crs_matrix->describe(*out, Teuchos::VERB_EXTREME);
+  #endif
+
+  timerGlobal = Teuchos::null;
+
+  // Save crs_matrix as a MatrixMarket file.
+  std::ofstream ofs("Finite-Element-Matrix-Assembly_Type3.out", std::ofstream::out);
+  Tpetra::MatrixMarket::Writer<MatrixType>::writeSparse(ofs, crs_matrix);
```

Comment:
Would be nice to control this via a command-line argument.

Reply:
I'll add that into the command line stuff when I put that in.  

Speaking of command-line processing and best-practices... should I use the [`Teuchos::CommandLineProcessor`][1] class or is something else preferred?

[1]: https://trilinos.org/docs/dev/packages/teuchos/doc/html/classTeuchos_1_1CommandLineProcessor.html

Reply:
`Teuchos::CommandLineProcessor` is fine for now.